extension-manifest-v3/src/background/reporting.js

```diff
@@ -216,26 +216,45 @@ const reporting = new Reporting({
   communication,
 });
 
-observe('terms', (terms) => {
+function delay(timeInMs) {
+  return new Promise((resolve) => setTimeout(resolve, timeInMs));
+}
+
+function defer() {
+  const deferred = {};
+  deferred.promise = new Promise((resolver) => {
+    deferred.resolve = resolver;
+  });
+  deferred.waitUntil = async ({
+    timeout,
+    errorMessage = 'Timeout exceeded',
+  } = {}) =>
+    Promise.race([
+      deferred.promise,
+      delay(timeout).then(() => Promise.reject(errorMessage)),
+    ]);
+  return deferred;
+}
+
+const deferredStartup = defer();
+
+observe('terms', async (terms) => {
   if (terms) {
-    reporting.init().catch((e) => {
+    try {
+      await reporting.init();
+    } catch (e) {
       console.warn(
         'Failed to initialize reporting. Leaving the module disabled and continue.',
         e,
       );
-    });
+    }
   } else {
     reporting.unload();
   }
+  deferredStartup.resolve(reporting.isActive);
```

Comment:
Combining the state and a defer is problematic, since it won't detect changes.

For instance:
* Start with a new installation (the deferred will soon resolve to "false")
* Enable terms (the module now initializes, but the deferred is still "false")

In the end, the module is technically enabled, but because of the "if" below it remains practically disabled.

Reply:
I see.

Isn't the same happening in the other implementation. In the and pendingPromise was being used only one time.

Reply:
In https://github.com/ghostery/ghostery-extension/pull/1115/commits/1a5a373f6e91b820d95613f5c89ab5f9d3836280 or https://github.com/ghostery/ghostery-extension/pull/1115/files, the goal of pendingStartup is to delay the processing of the event. It doesn't have the semantic that the module is afterwards fully loaded. Also, it relies on the initialization in the module itself.

The main challenge is that to initialize, you need to read persisted state (to learn if it starts enabled or disabled). That forces at least one async operation. That unfortunately clashes with the other requirement in service workers to be able to process events on the first tick.

(That is where the idea to delay comes from. It's not the only implementation strategy; buffering events during startup would be another. Though buffering may conflict with in low-latency requirements, e.g. the adblocker.)